File: src/main/java/xyz/matthewtgm/tgmlib/TGMLibInstaller.java
package xyz.matthewtgm.tgmlib;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import net.minecraft.launchwrapper.Launch;
import net.minecraft.launchwrapper.LaunchClassLoader;
import org.apache.commons.io.IOUtils;

import javax.swing.*;
import java.awt.*;
import java.io.*;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

public class TGMLibInstaller {

    private static final Gson gson = new Gson();

    private static final String versions_url = "https://raw.githubusercontent.com/TGMDevelopment/TGMLib-Data/main/versions.json";
    private static final String className = "xyz.matthewtgm.tgmlib.TGMLib";
    private static boolean initialized;
    private static File dataDir;

    public static boolean isInitialized() {
        try {
            Launch.classLoader.clearNegativeEntries(new HashSet<>(Collections.singletonList(className)));
            Field invalidClasses = LaunchClassLoader.class.getDeclaredField("invalidClasses");
            if (!invalidClasses.isAccessible())
                invalidClasses.setAccessible(true);
            Object obj = invalidClasses.get(Launch.classLoader);
            ((Set<String>) obj).remove(className);
            return Class.forName(className) != null;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public static boolean isLoaded() {
        return initialized;
    }

    public static void load(File gameDir) {
        if (!isLoaded())
            return;

        try {
            Class<?> tgmLib = Class.forName(className);
            Method instance = tgmLib.getDeclaredMethod("getInstance");
            System.out.println(instance);
            Method initialize = tgmLib.getDeclaredMethod("initialize", File.class);
            System.out.println(initialize);
            initialize.invoke(instance.invoke(null), gameDir);
            System.out.println("Loaded TGMLib successfully.");
        } catch (Exception e) {
            e.printStackTrace();
        }
        System.out.println("Did NOT load TGMLib successfully.");
    }

    public static ReturnValue initialize(File gameDir) {
        if (isInitialized())
            return ReturnValue.ALREADY_INITIALIZED;

        dataDir = new File(gameDir, "TGMLib");
        if (!dataDir.exists() || !dataDir.isDirectory())
            if (!dataDir.mkdirs())
                return ReturnValue.FAILED;

        JsonObject versionsJson = gson.fromJson(fetchJson(versions_url), JsonObject.class);

        File metaFile = new File(dataDir, "meta.json");

        if (!metaFile.exists() || !metaFile.isFile())
            createAndUpdateMetaFile(metaFile, versionsJson);

        JsonObject localMetadataJson = gson.fromJson(readJsonFromFile(new File(dataDir, "meta.json")), JsonObject.class);

        File tgmLibFile = new File(dataDir, "TGMLib-" + versionsJson.get("latest").getAsString() + ".jar");

        if (!tgmLibFile.exists()) {
            File old = new File(dataDir, "TGMLib-" + localMetadataJson.get("current").getAsString() + ".jar");
            if (old.exists())
                if (!old.delete())
                    return ReturnValue.FAILED;

            if (!download("https://raw.githubusercontent.com/TGMDevelopment/TGMLib-Data/main/downloads/TGMLib-" + versionsJson.get("latest").getAsString() + ".jar", versionsJson.get("latest").getAsString(), tgmLibFile))
                return ReturnValue.FAILED;
        }

        createAndUpdateMetaFile(metaFile, versionsJson);

        addToClasspath(tgmLibFile);

        initialized = true;
        if (!isInitialized())
            return ReturnValue.FAILED;
        return ReturnValue.SUCCESSFUL;
    }

    private static void createAndUpdateMetaFile(File metaFile, JsonObject versionsJson) {
        BufferedWriter writer = null;
        try {
            if (!metaFile.exists() || !metaFile.isFile())
                metaFile.createNewFile();
            writer = new BufferedWriter(new FileWriter(metaFile));
            writer.write("{\"current\": \"" + versionsJson.get("latest").getAsString() + "\"}");
        } catch (IOException e) {
            e.printStackTrace();
            try {
                writer.flush();
                writer.close();
            } catch (Exception e1) {
                e1.printStackTrace();
            }
        } finally {
            try {
                writer.flush();
                writer.close();
            } catch (Exception e1) {
                e1.printStackTrace();
            }
        }
    }

    private static boolean download(String url, String version, File file) {
        url = url.replace(" ", "%20");

        System.out.println("Downloading TGMLib...\n(URL: " + url + ")");

        try {
            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
        } catch (Exception e) {
            e.printStackTrace();
        }

        JFrame frame = new JFrame("TGMLib Installer");
        JProgressBar progressBar = new JProgressBar();
        JLabel label = new JLabel("Downloading TGMLib " + version, SwingConstants.CENTER);
        label.setSize(600, 120);
        frame.getContentPane().add(label);
        frame.getContentPane().add(progressBar);
        GroupLayout layout = new GroupLayout(frame.getContentPane());
        frame.getContentPane().setLayout(layout);
        layout.setHorizontalGroup(layout.createParallelGroup(GroupLayout.Alignment.LEADING)
                .addGroup(layout.createSequentialGroup()
                        .addContainerGap()
                        .addGroup(layout.createParallelGroup(GroupLayout.Alignment.LEADING)
                                .addComponent(label, GroupLayout.Alignment.TRAILING, GroupLayout.DEFAULT_SIZE, 376, Short.MAX_VALUE)
                                .addComponent(progressBar, GroupLayout.Alignment.TRAILING, GroupLayout.DEFAULT_SIZE, GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
                        .addContainerGap()));
        layout.setVerticalGroup(layout.createParallelGroup(GroupLayout.Alignment.LEADING)
                .addGroup(layout.createSequentialGroup()
                        .addContainerGap()
                        .addComponent(label, GroupLayout.PREFERRED_SIZE, 55, GroupLayout.PREFERRED_SIZE)
                        .addPreferredGap(LayoutStyle.ComponentPlacement.RELATED)
                        .addComponent(progressBar, GroupLayout.PREFERRED_SIZE, 33, GroupLayout.PREFERRED_SIZE)
                        .addContainerGap(GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)));
        frame.setResizable(false);
        progressBar.setBorderPainted(true);
        progressBar.setMinimum(0);
        progressBar.setStringPainted(true);
        Font font = progressBar.getFont();
        progressBar.setFont(new Font(font.getName(), font.getStyle(), font.getSize() * 2));
        label.setFont(new Font(font.getName(), font.getStyle(), font.getSize() * 2));
        frame.pack();
        frame.setLocationRelativeTo(null);
        frame.setVisible(true);

        HttpURLConnection connection = null;
        try (FileOutputStream fout = new FileOutputStream(file)) {
            URL theUrl = new URL(url);
            connection = (HttpURLConnection) theUrl.openConnection();
            connection.setRequestMethod("GET");
            connection.setUseCaches(true);
            connection.setReadTimeout(20000);
            connection.setConnectTimeout(15000);
            connection.setDoOutput(true);

            int contentLength = connection.getContentLength();
            byte[] buffer = new byte[1024];

            progressBar.setMaximum(contentLength);

            int read;
            progressBar.setValue(0);
            while ((read = connection.getInputStream().read(buffer)) > 0) {
                fout.write(buffer, 0, read);
                progressBar.setValue(progressBar.getValue() + 1024);
            }
        } catch (Exception e) {
            if (!(e instanceof UnsupportedOperationException))
                e.printStackTrace();
            frame.dispose();
            return false;
        } finally {
            if (connection != null)
                connection.disconnect();
        }

        System.out.println("Finished downloading TGMLib!");

        frame.dispose();
        return true;
    }

    private static void addToClasspath(File file) {
        try {
            URL url = file.toURI().toURL();
            System.out.println(url);

            ClassLoader classLoader = TGMLibInstaller.class.getClassLoader();
            Method method = classLoader.getClass().getDeclaredMethod("addURL", URL.class);
            method.setAccessible(true);
            method.invoke(classLoader, url);
        } catch (Exception e) {
            throw new RuntimeException("Unexpected exception...", e);
        }
    }

    private static String readJsonFromFile(File file) {
        try {
            StringBuilder builder = new StringBuilder();
            BufferedReader reader = new BufferedReader(new FileReader(file));
            reader.lines().forEach(builder::append);
            return builder.toString();
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    private static String fetchJson(String url) {
        url = url.replace(" ", "%20");
        HttpURLConnection connection = null;
        try {
            URL theUrl = new URL(url);
            connection = (HttpURLConnection) theUrl.openConnection();
            connection.setRequestMethod("GET");
            connection.setUseCaches(true);
            connection.setReadTimeout(20000);
            connection.setConnectTimeout(15000);
            connection.setDoOutput(true);
            return IOUtils.toString(connection.getInputStream(), Charset.defaultCharset());
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        } finally {
            if (connection != null)
                connection.disconnect();
        }
    }

    public enum ReturnValue {
        ALREADY_INITIALIZED,
        FAILED,
        ERRORED,
        SUCCESSFUL
    }

}

File: src/main/kotlin/xyz/matthewtgm/lib/TGMLibInstaller.kt
package xyz.matthewtgm.lib

import com.google.gson.Gson
import com.google.gson.JsonObject
import net.minecraft.launchwrapper.Launch
import net.minecraft.launchwrapper.LaunchClassLoader
import org.apache.commons.io.IOUtils
import java.awt.Font
import java.io.*
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.Charset
import java.util.*
import javax.swing.*


object TGMLibInstaller {
    private val gson = Gson()
    private const val versions_url = "https://raw.githubusercontent.com/TGMDevelopment/TGMLib-Data/main/versions.json"
    private const val className = "xyz.matthewtgm.tgmlib.TGMLib"
    var isLoaded = false
        private set
    private var dataDir: File? = null
    fun isInitialized(): Boolean {
        try {
            Launch.classLoader.clearNegativeEntries(HashSet(Collections.singletonList(className)))
            val invalidClasses = LaunchClassLoader::class.java.getDeclaredField("invalidClasses")
            if (!invalidClasses.isAccessible) invalidClasses.isAccessible = true
            val obj = invalidClasses[Launch.classLoader]
            (obj as MutableSet<*>).remove(className)
            return Class.forName(className) != null
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    fun load(gameDir: File?) {
        if (!isLoaded) return
        try {
            val tgmLib = Class.forName(className)
            val instance = tgmLib.getDeclaredMethod("getInstance")
            println(instance)
            val initialize = tgmLib.getDeclaredMethod("initialize", File::class.java)
            println(initialize)
            initialize.invoke(instance.invoke(null), gameDir)
            println("Loaded TGMLib successfully.")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        println("Did NOT load TGMLib successfully.")
    }

    fun initialize(gameDir: File?): ReturnValue {
        if (isInitialized()) return ReturnValue.ALREADY_INITIALIZED
        dataDir = File(gameDir, "TGMLib")
        if (!dataDir!!.exists() || !dataDir!!.isDirectory) if (!dataDir!!.mkdirs()) return ReturnValue.FAILED
        val versionsJson = gson.fromJson(
            fetchJson(versions_url),
            JsonObject::class.java
        )
        val metaFile = File(dataDir, "meta.json")
        if (!metaFile.exists() || !metaFile.isFile) createAndUpdateMetaFile(metaFile, versionsJson)
        val localMetadataJson = gson.fromJson(
            readJsonFromFile(File(dataDir, "meta.json")),
            JsonObject::class.java
        )
        val tgmLibFile = File(dataDir, "TGMLib-" + versionsJson["latest"].asString + ".jar")
        if (!tgmLibFile.exists()) {
            val old = File(dataDir, "TGMLib-" + localMetadataJson["current"].asString + ".jar")
            if (old.exists()) if (!old.delete()) return ReturnValue.FAILED
            if (!download(
                    "https://raw.githubusercontent.com/TGMDevelopment/TGMLib-Data/main/downloads/TGMLib-" + versionsJson["latest"].asString + ".jar",
                    versionsJson["latest"].asString,
                    tgmLibFile
                )
            ) return ReturnValue.FAILED
        }
        createAndUpdateMetaFile(metaFile, versionsJson)
        addToClasspath(tgmLibFile)
        isLoaded = true
        return if (!isInitialized()) ReturnValue.FAILED else ReturnValue.SUCCESSFUL
    }

    private fun createAndUpdateMetaFile(metaFile: File, versionsJson: JsonObject) {
        var writer: BufferedWriter? = null
        try {
            if (!metaFile.exists() || !metaFile.isFile) metaFile.createNewFile()
            writer = BufferedWriter(FileWriter(metaFile))
            writer.write("{\"current\": \"" + versionsJson["latest"].asString + "\"}")
        } catch (e: IOException) {
            e.printStackTrace()
            try {
                writer!!.flush()
                writer.close()
            } catch (e1: Exception) {
                e1.printStackTrace()
            }
        } finally {
            try {
                writer!!.flush()
                writer.close()
            } catch (e1: Exception) {
                e1.printStackTrace()
            }
        }
    }

    private fun download(url: String, version: String, file: File): Boolean {
        var url = url
        url = url.replace(" ", "%20")
        println("Downloading TGMLib...\n(URL: $url)")
        try {
            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName())
        } catch (e: Exception) {
            e.printStackTrace()
        }
        val frame = JFrame("TGMLib Installer")
        val progressBar = JProgressBar()
        val label = JLabel("Downloading TGMLib $version", SwingConstants.CENTER)
        label.setSize(600, 120)
        frame.contentPane.add(label)
        frame.contentPane.add(progressBar)
        val layout = GroupLayout(frame.contentPane)
        frame.contentPane.layout = layout
        layout.setHorizontalGroup(
            layout.createParallelGroup(GroupLayout.Alignment.LEADING)
                .addGroup(
                    layout.createSequentialGroup()
                        .addContainerGap()
                        .addGroup(
                            layout.createParallelGroup(GroupLayout.Alignment.LEADING)
                                .addComponent(
                                    label,
                                    GroupLayout.Alignment.TRAILING,
                                    GroupLayout.DEFAULT_SIZE,
                                    376,
                                    Short.MAX_VALUE.toInt()
                                )
                                .addComponent(
                                    progressBar,
                                    GroupLayout.Alignment.TRAILING,
                                    GroupLayout.DEFAULT_SIZE,
                                    GroupLayout.DEFAULT_SIZE,
                                    Short.MAX_VALUE.toInt()
                                )
                        )
                        .addContainerGap()
                )
        )
        layout.setVerticalGroup(
            layout.createParallelGroup(GroupLayout.Alignment.LEADING)
                .addGroup(
                    layout.createSequentialGroup()
                        .addContainerGap()
                        .addComponent(label, GroupLayout.PREFERRED_SIZE, 55, GroupLayout.PREFERRED_SIZE)
                        .addPreferredGap(LayoutStyle.ComponentPlacement.RELATED)
                        .addComponent(progressBar, GroupLayout.PREFERRED_SIZE, 33, GroupLayout.PREFERRED_SIZE)
                        .addContainerGap(GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE.toInt())
                )
        )
        frame.isResizable = false
        progressBar.isBorderPainted = true
        progressBar.minimum = 0
        progressBar.isStringPainted = true
        val font = progressBar.font
        progressBar.font = Font(font.name, font.style, font.size * 2)
        label.font = Font(font.name, font.style, font.size * 2)
        frame.pack()
        frame.setLocationRelativeTo(null)
        frame.isVisible = true
        var connection: HttpURLConnection? = null
        try {
            FileOutputStream(file).use { fout ->
                val theUrl = URL(url)
                connection = theUrl.openConnection() as HttpURLConnection
                connection!!.requestMethod = "GET"
                connection!!.useCaches = true
                connection!!.readTimeout = 20000
                connection!!.connectTimeout = 15000
                connection!!.doOutput = true
                val contentLength = connection!!.contentLength
                val buffer = ByteArray(1024)
                progressBar.maximum = contentLength
                var read: Int
                progressBar.value = 0
                while (connection!!.inputStream.read(buffer).also { read = it } > 0) {
                    fout.write(buffer, 0, read)
                    progressBar.value = progressBar.value + 1024
                }
            }
        } catch (e: Exception) {
            if (e !is UnsupportedOperationException) e.printStackTrace()
            frame.dispose()
            return false
        } finally {
            if (connection != null) connection!!.disconnect()
        }
        println("Finished downloading TGMLib!")
        frame.dispose()
        return true
    }

    private fun addToClasspath(file: File) {
        try {
            val url = file.toURI().toURL()
            println(url)
            val classLoader = TGMLibInstaller::class.java.classLoader
            val method = classLoader.javaClass.getDeclaredMethod("addURL", URL::class.java)
            method.isAccessible = true
            method.invoke(classLoader, url)
        } catch (e: Exception) {
            throw RuntimeException("Unexpected exception...", e)
        }
    }

    private fun readJsonFromFile(file: File): String? {
        return try {
            val builder = StringBuilder()
            val reader = BufferedReader(FileReader(file))
            reader.lines().forEach { str: String? ->
                builder.append(
                    str
                )
            }
            builder.toString()
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    private fun fetchJson(url: String): String? {
        var url = url
        url = url.replace(" ", "%20")
        var connection: HttpURLConnection? = null
        return try {
            val theUrl = URL(url)
            connection = theUrl.openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.useCaches = true
            connection.readTimeout = 20000
            connection.connectTimeout = 15000
            connection.doOutput = true
            IOUtils.toString(connection.inputStream, Charset.defaultCharset())
        } catch (e: Exception) {
            e.printStackTrace()
            null
        } finally {
            connection?.disconnect()
        }
    }

    enum class ReturnValue {
        ALREADY_INITIALIZED, FAILED, ERRORED, SUCCESSFUL
    }
}